libultraship/libultraship/BinaryReader.cpp

```diff
@@ -0,0 +1,211 @@
+#include "BinaryReader.h"
+#include <cmath>
+#include "Lib/BinaryTools/BinaryTools/BinaryReader.h"
+
+Ship::BinaryReader::BinaryReader(char* buffer, uint32_t size) 
+{
+    this->buffer = std::vector<char>(buffer, buffer + size);
+    this->reader = std::make_shared<::BinaryReader>(this->buffer.data(), size);
+}
+
+void Ship::BinaryReader::SetEndianness(Endianness endianness) 
+{
+    this->endianness = endianness;
+}
+
+Ship::Endianness Ship::BinaryReader::GetEndianness() const
+{
+    return endianness;
+}
+
+void Ship::BinaryReader::Seek(int32_t offset, SeekOffsetType seekType) {
+    switch(seekType) {
+        case SeekOffsetType::Current:
+            if (offset < 0) {
+                reader->SeekReverse(-1 * offset);
+                break;
+            }
+            reader->SeekCur(offset);
+            break;
+        case SeekOffsetType::Start:
+            reader->SeekBeg(offset);
+            break;
+        case SeekOffsetType::End:
+            reader->SeekReverse(offset);
+            break;
+    }
+}
+
+uint32_t Ship::BinaryReader::GetBaseAddress() {
+    return reader->Position();
+}
+
+void Ship::BinaryReader::Read(int32_t length)
+{
+    reader->ReadFixedLengthString(length);
+}
+
+void Ship::BinaryReader::Read(char *buffer, int32_t length)
+{
+    reader->ReadToMemory(buffer, length);
+}
+
+char Ship::BinaryReader::ReadChar()
+{
+    return reader->ReadChar();
+}
+
+int8_t Ship::BinaryReader::ReadByte()
+{
+    return reader->ReadInt8();
+}
+
+int16_t Ship::BinaryReader::ReadInt16()
+{
+    int16_t result = 0;
+    this->Read((char*)&result, sizeof(int16_t));
+    if (endianness != Endianness::Native)
+        result = BSWAP16(result);
+
+    return result;
+}
+
+int32_t Ship::BinaryReader::ReadInt32()
+{
+    int32_t result = 0;
+
+    this->Read((char *)&result, sizeof(int32_t));
+
+    if (endianness != Endianness::Native)
+        result = BSWAP32(result);
+
+    return result;
+}
+
+uint8_t Ship::BinaryReader::ReadUByte()
+{
+    return reader->ReadUint8();
+}
+
+uint16_t Ship::BinaryReader::ReadUInt16()
+{
+    uint16_t result = 0;
+
+    this->Read((char *)&result, sizeof(uint16_t));
+
+    if (endianness != Endianness::Native)
+        result = BSWAP16(result);
+
+    return result;
+}
+
+uint32_t Ship::BinaryReader::ReadUInt32()
+{
+    uint32_t result = 0;
+
+    this->Read((char *)&result, sizeof(uint32_t));
+
+    if (endianness != Endianness::Native)
+        result = BSWAP32(result);
+
+    return result;
+}
+
+uint64_t Ship::BinaryReader::ReadUInt64()
+{
+    uint64_t result = 0;
+
+    this->Read((char *)&result, sizeof(uint64_t));
+
+    if (endianness != Endianness::Native)
+        result = BSWAP64(result);
+
+    return result;
+}
+
+float Ship::BinaryReader::ReadSingle()
```

Comment:
```suggestion
float Ship::BinaryReader::ReadFloat()
```

Reply:
I can change that but I did it that way because I was having my wrapper mimic the previous API which had ReadSingle instead of Float for some reason.

Reply:
For this and the Byte->Int8 case, do you want me to go through and change all the calls to ReadSingle that were already there? Or just add the new function definition so that it can be used in the future?

Reply:
Normally singles are called floats and int8 and uint8 is more consistent, so I say we update everything to match that.

Reply:
Alright, most recent commit should do this for Floats and Int8s. I hadn't changed UByte to UInt8 yet but I'll do that as well in a minute.